lib/pbench/agent/results.py

```diff
@@ -165,7 +165,7 @@ def __init__(self, tarball, config, logger):
         self.tarball_md5 = Path(tarball_md5)
         self.logger = logger
         server_rest_url = config.results.get("server_rest_url")
-        self.upload_url = f"{server_rest_url}/upload"
+        self.upload_url = f"{server_rest_url}/upload/ctrl/controller"
```

Comment:
This appears to be a "fixed", or hard-coded, controller named, `controller`.

The current `bash` implementation starts in `pbench-move-results`, where at [line 85](https://github.com/distributed-system-analysis/pbench/blob/master/agent/util-scripts/pbench-move-results#L85), it sets the `controller` to be the `full_hostname` environment variable, as provided by the [`base`](https://github.com/distributed-system-analysis/pbench/blob/master/agent/base#L128) script.  Then `pbench-move-results` constructs a temporary directory at [line 172](https://github.com/distributed-system-analysis/pbench/blob/master/agent/util-scripts/pbench-move-results#L172), which is where it places the constructed tar ball from `pbench-make-result-tb` (see [line 198](https://github.com/distributed-system-analysis/pbench/blob/master/agent/util-scripts/pbench-move-results#L198)).

The old code did it that way so that it could use one (1) `scp` command (see [line 68](https://github.com/distributed-system-analysis/pbench/blob/master/agent/util-scripts/pbench-copy-result-tb#L68) of `pbench-copy-result-tb`), followed by one (1) `ssh` command to verify the bits on the far side.  This sequence allowed us to copy the `.tar.xz` and the `.tar.xz.md5` file in one `scp` command, also providing the `controller` name (the name of the directory in which the tar ball is contained) without making multiple calls to `scp` or `ssh` beyond the two required (copy the data, verify the MD5 sum holds).

In this new HTTP/S PUT API, we place the `controller` in the URI, the expected MD5 checksum value in the `Content-MD5` header, and `.tar.xz` contents in the payload body.

We almost have all the pieces in place, except that a `controller` needs to be placed into the URI the client forms.

It looks like we would want to use the `controller` found at [line 14](https://github.com/distributed-system-analysis/pbench/blob/master/lib/pbench/cli/agent/commands/results/__init__.py#L14) of the new Python3-based `pbench move/copy results` command(s), and pass that into the `CopyResultTb()` constructor so that it can form the URI properly at [line 168](https://github.com/distributed-system-analysis/pbench/blob/master/lib/pbench/agent/results.py#L168).

Reply:
Thanks @portante for the detailed explanation. It make sense to update the constructor of the `CopyResultTb` to take `controller` name since its client defined now.